app/scripts/controller-init/confirmations/transaction-controller-init.ts

```diff
@@ -247,6 +260,62 @@ function publishSmartTransactionHook(
   });
 }
 
+function publishBatchSmartTransactionHook({
+  transactionController,
+  smartTransactionsController,
+  hookControllerMessenger,
+  flatState,
+  transactions,
+}: {
+  transactionController: TransactionController;
+  smartTransactionsController: SmartTransactionsController;
+  hookControllerMessenger: SmartTransactionHookMessenger;
+  flatState: ControllerFlatState;
+  transactions: PublishBatchHookTransaction[];
+}) {
+  // UI state is required to support shared selectors to avoid duplicate logic in frontend and backend.
+  // Ideally all backend logic would instead rely on messenger event / state subscriptions.
+  const uiState = getUIState(flatState);
+
+  // @ts-expect-error Smart transaction selector types does not match controller state
+  const isSmartTransaction = getIsSmartTransaction(uiState);
+
+  if (!isSmartTransaction) {
+    // Will cause TransactionController to publish to the RPC provider as normal.
+    return undefined;
+  }
+
+  // @ts-expect-error Smart transaction selector types does not match controller state
+  const featureFlags = getFeatureFlagsByChainId(uiState);
+
+  // Get transactionMeta based on the last transaction ID
+  const lastTransaction = transactions[transactions.length - 1];
```

Comment:
Could this logic be encapsulated inside the other file and we just pass in the ID?

Reply:
The hook constructor expects `transactionMeta`, so I want to get it here before I call `submitBatchSmartTransactionHook`. I also want to throw an error early if there is no `transactionMeta`, so I think keeping it here makes sense.